pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,442 @@
+				log::error!("Failed to fetch price from oracle with error: {}", e);
+				return Err(e)
+			}
+			Ok(())
+		}
+
+		fn fetch_price(
+			block_number: T::BlockNumber,
+			contract: &str,
+			geth_uri: &str
+		) -> Result<(), &'static str> {
+			let req = json!({
+				"jsonrpc": "2.0",
+				"method": "eth_call", // https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_call
+				"params": [
+					{
+					"to": contract,
+					// first 4 bytes of keccak_256(latestRoundData()) function, padded - Use https://emn178.github.io/online-tools/keccak_256.html
+					"data": "feaf968c0000000000000000000000000000000000000000000000000000000000000000",
+					},
+					"latest"
+				],
+				 "id": "5", //goerli
+			});
+
+			let req = serde_json::to_string(&req).map_err(|_| "Invalid request")?;
+			log::trace!("Request: {}", req);
+
+			let response_body = http_json_post(&geth_uri, req.as_bytes()); // Get the latest block number of the Ethereum Blockchain
+			let response_body = if let Ok(response) = response_body {
+				response
+			} else {
+				return Err("Failed to get response from geth")
+			};
+
+			let response = String::from_utf8(response_body).map_err(|_| "Invalid response")?;
+			log::trace!("Response: {}", response);
+
+			let v: Value =
+				serde_json::from_str(&response).map_err(|_| "Invalid response - json parse")?;
+			let result = v["result"].as_str().ok_or("Invalid response - no result")?; // Get the latest block number of the Ethereum Blockchain
+			let data = hex::decode(&result[2..]).map_err(|_| "Invalid response - invalid data")?;
+			let data = ethabi::decode(
+				//https://docs.chain.link/docs/data-feeds/price-feeds/api-reference/#latestrounddata
+				&[ParamType::Tuple(vec![
+					ParamType::Uint(80), // uint80 roundId
+					ParamType::Int(256), // int256 answer
+					ParamType::Uint(256), // uint256 startedAt
+					ParamType::Uint(256), // uint256 updatedAt
+					ParamType::Uint(80), // uint80 answeredInRound
+
+				])],
+				&data).map_err(|_| "Invalid response")?;
+			let new_price = data[1].clone().into_int().ok_or_else(|| "Invalid price").unwrap().as_u32();
+			log::trace!("New price: {}", new_price);
+
+			// -- Sign using any account
+			let (_, result) = Signer::<T, T::AuthorityId>::any_account()
+				.send_unsigned_transaction(
+					|account| PricePayload {
+						block_number,
+						price: new_price.into(),
+						public: account.public.clone()
+					},
+					|payload, signature| Call::store_price {
+						price_payload: payload,
+						signature,
+					},
+				)
+				.ok_or("No local accounts accounts available.")?;
+			result.map_err(|()| "Unable to submit transaction")?;
+
+			Ok(())
+		}
+
+		/// Add new price to the list.
+		fn add_price(block_number: T::BlockNumber, price: u32) {
+			log::info!("Adding to the average: {}", price);
+			<Prices<T>>::mutate(|prices| {
+				if prices.try_push(price).is_err() {
+					prices[(price % T::MaxPrices::get()) as usize] = price; // insert into the bounded size vector
```

Comment:
why do you do `(price % T::MaxPrices::get())` 😅?

Reply:
it's a bounded size vector so you need to remove one existing value to insert the new one. That is a way to do it, basically.

Reply:
So `(price % T::MaxPrices::get())` will give a random number from 0 to `T::MaxPrices::get() -1`, right?

But why do you want that number to be random 😅 ?

Why not make it a "bounded queue" instead of a "bounded vector"?

Reply:
"bounded queue" - so that after the queue reaches the length of `T::MaxPrices::get()` - the first element will  have to leave the queue?

(basically all existing elements get shifted by -1 index

Reply:
that also can work. How would you implement it?

Reply:
No idea 😅 😅 😅 . But that's a better solution right? Cos then we literally have the last `T::MaxPrices::get()` elements - since the first element always leaves rather than the element leaving is chosen at random  being random?

Actually, we can shift all elements behind by one index/position - but then the time complexity is `O(T::MaxPrices::get())`

Or actually I used `VecDeque` in `protos.get_geoneology()`- we can use that. And it seems there's also a bounded version: https://docs.rs/bounded-vec-deque/latest/bounded_vec_deque/struct.BoundedVecDeque.html

Reply:
In `Protos.get_genealogy` the order of items is important, so it makes sense to use a queue there. The complexity is O(N), which is also good. So queue is a good match.

In this case the order is not important, we just need the average, so the choice is between complexity O(1) (i.e. accessing the vector using index) and O(N). So, I think using vector is is not a bad choice.